server/proxy/pf_client.c

```diff
@@ -560,91 +774,126 @@ static int pf_logon_error_info(freerdp* instance, UINT32 data, UINT32 type)
+	WINPR_UNUSED(instance);
+
+	if (!pc)
+		return;
+
+	pc->sendChannelData = NULL;
+	ArrayList_Free(pc->cached_server_channel_data);
+	Stream_Free(pc->remote_pem, TRUE);
+	free(pc->remote_hostname);
 }
 
-/**
- * Callback set in the rdp_freerdp structure, and used to make a certificate validation
- * when a stored certificate does not match the remote counterpart.
- * This function will actually be called by tls_verify_certificate().
- * @see rdp_client_connect() and tls_connect()
- * @param instance        pointer to the rdp_freerdp structure that contains the connection settings
- * @param host            The host currently connecting to
- * @param port            The port currently connecting to
- * @param common_name     The common name of the certificate, should match host or an alias of it
- * @param subject         The subject of the certificate
- * @param issuer          The certificate issuer name
- * @param fingerprint     The fingerprint of the certificate
- * @param old_subject     The subject of the previous certificate
- * @param old_issuer      The previous certificate issuer name
- * @param old_fingerprint The fingerprint of the previous certificate
- * @param flags           See VERIFY_CERT_FLAG_* for possible values.
- *
- * @return 1 if the certificate is trusted, 2 if temporary trusted, 0 otherwise.
- */
-static DWORD pf_client_verify_changed_certificate_ex(
-    freerdp* instance, const char* host, UINT16 port, const char* common_name, const char* subject,
-    const char* issuer, const char* fingerprint, const char* old_subject, const char* old_issuer,
-    const char* old_fingerprint, DWORD flags)
+static int pf_client_verify_X509_certificate(freerdp* instance, const BYTE* data, size_t length,
+                                             const char* hostname, UINT16 port, DWORD flags)
 {
-	/* TODO: Add trust level to proxy configurable settings */
+	pClientContext* pc;
+
+	WINPR_ASSERT(instance);
+	WINPR_ASSERT(data);
+	WINPR_ASSERT(length > 0);
+	WINPR_ASSERT(hostname);
+
+	pc = (pClientContext*)instance->context;
+	WINPR_ASSERT(pc);
+
+	if (!Stream_EnsureCapacity(pc->remote_pem, length))
+		return 0;
+	Stream_SetPosition(pc->remote_pem, 0);
+
+	free(pc->remote_hostname);
+	pc->remote_hostname = NULL;
+
+	if (length > 0)
+		Stream_Write(pc->remote_pem, data, length);
+
+	if (hostname)
+		pc->remote_hostname = _strdup(hostname);
+	pc->remote_port = port;
+	pc->remote_flags = flags;
+
+	Stream_SealLength(pc->remote_pem);
+	if (!pf_modules_run_hook(pc->pdata->module, HOOK_TYPE_CLIENT_VERIFY_X509, pc->pdata, pc))
+		return 0;
 	return 1;
 }
 
-static void pf_client_context_free(freerdp* instance, rdpContext* context)
+static void* channel_data_copy(const void* obj)
 {
-	pClientContext* pc = (pClientContext*)context;
+	const proxyChannelDataEventInfo* src = obj;
+	proxyChannelDataEventInfo* dst;
 
-	if (!pc)
-		return;
+	WINPR_ASSERT(src);
+
+	dst = calloc(1, sizeof(proxyChannelDataEventInfo));
+	WINPR_ASSERT(dst);
```

Comment:
I don't think an assert makes sense here

Reply:
true, should abort.